internal/newusers/manager.go

```diff
@@ -0,0 +1,338 @@
+}
+
+// requestClearDatabase ask for the clean goroutine to clear up the database.
+// If we already have a pending request, do not block on it.
+// TODO: improve behavior when cleanup is already running
+// (either remove the dangling dirty file or queue the cleanup request).
+func (m *Manager) requestClearDatabase() {
+	if err := os.WriteFile(dirtyFlagDbName, nil, 0600); err != nil {
+		slog.Warn(fmt.Sprintf("Could not write dirty file flag to signal clearing up the database: %v", err))
+	}
+	select {
+	case m.doClear <- struct{}{}:
+	case <-time.After(10 * time.Millisecond): // Let the time for the cleanup goroutine for the initial start.
+	}
+}
+
+func (m *Manager) startUserCleanupRoutine(cacheDir string, opts *options) {
+	cleanupRoutineStarted := make(chan struct{})
+	go func() {
+		defer close(m.cleanupStopped)
+		close(cleanupRoutineStarted)
+		for {
+			select {
+			case <-m.doClear:
+				func() {
+					m.cache.ClearAndRebuild(cacheDir)
+				}()
+
+			case <-time.After(opts.cleanupInterval):
+				func() {
+					activeUsers, err := getActiveUsers(opts.procDir)
+					if err != nil {
+						slog.Warn(fmt.Sprintf("Could not get list of active users: %v", err))
+						return
+					}
+
+					if err := m.cache.CleanExpiredUsers(activeUsers, opts.expirationDate); err != nil {
+						slog.Warn(fmt.Sprintf("Could not clean database: %v", err))
+					}
+				}()
+
+			case <-m.quit:
+				return
+			}
+		}
+	}()
+	<-cleanupRoutineStarted
+}
+
+// getActiveUsers walks through procDir and returns a map with the usernames of the owners of all active processes.
+func getActiveUsers(procDir string) (activeUsers map[string]struct{}, err error) {
+	defer decorate.OnError(&err, "could not get list of active users")
+
+	activeUsers = make(map[string]struct{})
+
+	dirEntries, err := os.ReadDir(procDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, dirEntry := range dirEntries {
+		// Checks if the dirEntry represents a process dir (i.e. /proc/<pid>/)
+		if _, err := strconv.Atoi(dirEntry.Name()); err != nil {
+			continue
+		}
+
+		info, err := dirEntry.Info()
+		if err != nil {
+			// If the file doesn't exist, it means the process is not running anymore so we can ignore it.
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			}
+			return nil, err
+		}
+
+		stats, ok := info.Sys().(*syscall.Stat_t)
+		if !ok {
+			return nil, fmt.Errorf("could not get ownership of file %q", info.Name())
+		}
+
+		u, err := user.LookupId(strconv.Itoa(int(stats.Uid)))
+		if err != nil {
+			// Possibly a ghost/orphaned UID - no reason to error out. Just warn the user and continue.
+			slog.Warn(fmt.Sprintf("Could not map active user ID to an actual user: %v", err))
+			continue
+		}
+
+		activeUsers[u.Name] = struct{}{}
+	}
+	return activeUsers, nil
+}
+
+// getCache returns the underlying cache. This is only used for tests.
+//
+//nolint:unused // This is used for tests, with go linking. Not part of exported API.
+func (m *Manager) getCache() *cache.Cache {
```

Comment:
A small detail: as this is never used internally, I would directly import the struct in the `tests/` subpackage via linking and access the private field from `GetManagerCache`.

Reply:
I think `go:linkname` only works for functions and variables, no?

Reply:
Nope, I have done it in the past. Remember when I told you I did really really bad thing with the i18n before the changes were accepted upstream? Let me find it back…

Reply:
Sorry, my memory was wrong, I was using reflect (which you can there too) to get the field. As it’s only for reading it and not modifying the field itself (we can modify the object pointed by the field though), reflection sounds like the way to get there and avoid that linkname.